lib/Extension/LanguageServerPhpCsFixer/Model/PhpCsFixerProcess.php

```diff
@@ -0,0 +1,114 @@
+{
+    public const EXIT_SOME_FILES_INVALID = 4;
+    public const EXIT_FILES_NEEDS_FIXING = 8;
+
+    /**
+     * @param array<string,string> $env
+     */
+    public function __construct(
+        private string $binPath,
+        private LoggerInterface $logger,
+        private array $env = [],
+    ) {
+    }
+
+    /**
+     * @param  string[] $options
+     *
+     * @return Promise<string>
+     */
+    public function fix(string $content, array $options = []): Promise
+    {
+        return call(function () use ($content, $options) {
+            /** @var Process */
+            $process = yield $this->run('fix', ...[...$options, '-']);
+
+            $stdin = $process->getStdin();
+            $stdin->write($content);
+            $stdin->end();
+
+            $stdout = yield buffer($process->getStdout());
+            $exitCode = yield $process->join();
+
+            if ($exitCode !== 0
+                && $exitCode !== self::EXIT_SOME_FILES_INVALID
+                && $exitCode !== self::EXIT_FILES_NEEDS_FIXING
+                && $exitCode !== (self::EXIT_SOME_FILES_INVALID | self::EXIT_FILES_NEEDS_FIXING)
+            ) {
+                throw new PhpCsFixerError(
+                    $exitCode,
+                    $process->getCommand(),
+                    yield buffer($process->getStderr()),
+                    $stdout
+                );
+            }
+
+            return $stdout;
+        });
+    }
+
+    /**
+     * @param string[] $options
+     *
+     * @return Promise<string>
+     */
+    public function describe(string $rule, array $options = []): Promise
+    {
+        return call(function () use ($rule, $options) {
+            /** @var Process */
+            $process = yield $this->run('describe', ...[...$options, $rule]);
+
+            $stdout = yield buffer($process->getStdout());
+            $exitCode = yield $process->join();
+
+            if ($exitCode !== 0) {
+                throw new PhpCsFixerError(
+                    $exitCode,
+                    $process->getCommand(),
+                    yield buffer($process->getStderr()),
+                    $stdout
+                );
+            }
+
+            return $stdout;
+        });
+    }
+
+    /**
+     * @return Promise<Process>
+     */
+    public function run(string ...$args): Promise
+    {
+        return call(function () use ($args) {
+            $process = new Process([$this->binPath, ...$args], null, $this->env);
+            yield $process->start();
+
+            $process->join()
+                ->onResolve(function (?Throwable $error, $data) use ($process): void {
+                    $this->logger->log(
+                        $error ? 'warning' : 'debug',
+                        sprintf(
+                            'Executed %s, which exited with %s',
+                            $process->getCommand(),
+                            $data
+                        )
+                    );
+                });
```

Comment:
any reason to use `onResolve`?

can do:

```
yield $process->join();
```
?

Reply:
I'm not going to lie, but I did forgot a bit of a implementation details and reasoning, so the explanation may be wrong. I'm about to leave now, can do little testing tommorow to find out whether what I write below is true.

However, I believe the reason is in lines 39-41, that write to `stdin` of the process. `onResolve` is used to add a listener for a logging purposes.
I believe if I used `yield $process->join();` with the process that expects `stdin` then that Promise would get stuck waiting for input. Input from lines 39-41 wouldn't happen, because we'd be still on line 37 waiting for run to complete.